### PR TITLE
Newsletters for change 

### PR DESCRIPTION
**Pull request by leonkoech**
## What does the PR do?

Created a Newsletter front end and API consumption for OHACK for opportunity Hack 2022. 
Team: `Newsletter for change`

Also created a template for pull requests like this

### Type of change

- [ ] Breaking Change
- [ ] Bug Fix
- [x] New Feature

## Linked Issue
Related Issue: [Enable Other Volunteer Oportunities](https://github.com/opportunity-hack/frontend-ohack.dev/issues/9)

## Make sure you have

- [x] Pulled from the default branch
- [x] Documented your changes
- [x] Linked the Issue
- [x] Appointed a reviewer (if any)

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+## Linked Issue
+Related Issue: [reference to a related issue](https://docs.github.com/en/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)
+
+## Make sure you have
+
+- [ ] Pulled from the default branch
+- [ ] Documented your changes
+- [ ] Linked the Issue
+- [ ] Appointed a reviewer (if any)
```
Comment (leonkoech) [inline .github/pull_request_template.md:17]: Git PR template :)

File: package.json [hunk 2]
```diff
@@ -19,18 +19,25 @@
     "@testing-library/react": "^11.1.0",
     "@testing-library/user-event": "^12.1.10",
     "axios": "^0.27.2",
+    "html-react-parser": "^3.0.8",
```
Comment (leonkoech) [inline package.json:22]: For the future preview page. Reference will be somewhere at the bottom. 

ATM an api returns the html page to be sent as a string, but has a few errors :'(

File: package.json [hunk 3]
```diff
@@ -19,18 +19,25 @@
...
+    "localforage": "^1.10.0",
+    "match-sorter": "^6.3.1",
     "eslint-config-react-app": "^7.0.1",
     "hamburger-react": "^2.5.0",
     "moment": "^2.29.4",
     "next": "^12.3.1",
     "react": "^17.0.2",
     "react-circular-progressbar": "^2.1.0",
     "react-dom": "^17.0.2",
+    "react-hot-toast": "^2.4.0",
```
Comment (leonkoech) [inline package.json:32]: For the nice toast floating above when an action happens

File: src/components/buttons/newsletter-button.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  const success = (value) => toast.success(value);
+  return (
+    <button
+      className="button dark-text locked light-text"
+      onClick={(e) => {
+        
+        check_subscription_status(props.user_id).then((subscribed) => {
+          if(subscribed == "Login required"){
+            error("Login Required")
+          }
```
Comment (leonkoech) [inline src/components/buttons/newsletter-button.js:19]: Sometimes the return value from `check_subscription_status` status is the error string, esp when the user hasn't logged in before subbing or unsubbing.

File: src/components/newsletter/custom_button.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            'button_text': valueOrUndefined(e.target.button_text.value)
+        }
+        console.log(new_values)
+        for(let [key,value] of Object.entries(new_values)){
+            if(value === undefined){
+                notify(key.split("_").join(" "), "is empty")
+            }
+        }
+        const final_text = '${"type":"'+new_values.button_type+'", "link":"'+new_values.url_link+'", "text":"'+new_values.button_text+'"}$'
+        navigator.clipboard.writeText(final_text)
```
Comment (leonkoech) [inline src/components/newsletter/custom_button.js:30]: Byron wanted to see this piece of code :)

File: src/components/newsletter/preview.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+export default function HTMLPreview(props){
+    return(
+        <html>
+            {
+                 parse(props.html)
+             }
+        </html>
+    )
+}
```
Comment (leonkoech) [inline src/components/newsletter/preview.js:11]: Intended for previewing the newsletter before sending :)

Comment (leonkoech) [reply src/components/newsletter/preview.js:11]: should work with simple texts as the body :)

File: src/components/newsletter/selected_users.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
+export default function SubscriberList (props){
+    const value = []
+    
+    if(props.subscribers != undefined) {
+      props.subscribers[props.item_name]?.forEach((v) => {
+        // filter here
+        if(String(v.subscribe) == props.is_subscribed){
```
Comment (leonkoech) [inline src/components/newsletter/selected_users.js:7]: filtering the email list at the front end instead of the backend as discussed on 02/03

File: src/hooks/use-newsletter-subscription.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+      } catch (error) {
+
+          if (axios.isAxiosError(error) && error.response) {
+              return error.response;
+          }
+
+          return error.message;
+      }
+  }, [getAccessTokenSilently]);
```
Comment (leonkoech) [inline src/hooks/use-newsletter-subscription.js:39]: to make requests to the API

Kept getting Axios errors at the beginning :'((

File: src/hooks/use-newsletter-subscription.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+      console.log(  `${apiServerUrl}/api/newsletter-subs/${subscription_state}/${user_id}`, data)
+      var new_data = data.data
+      new_data["subscribed"] =  new_data["subscribed"] == "true"
+      console.log(new_data)
+      return new_data
+    } else {
+      return undefined
+    }
+  }
+
```
Comment (leonkoech) [inline src/hooks/use-newsletter-subscription.js:65]: Subscription API HOOK 

@leonkoech remove the `console.log()`

File: src/hooks/use-newsletter-subscription.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    if (data) {
+      
+      var new_data = data.data
+      console.log(new_data["is_subscribed"] , "true")
+      return new_data["is_subscribed"] == "true"
+      // TODO: return boolean to stop loader if any
+    } else {
+      console.log("error");
+    }
+  }
```
Comment (leonkoech) [inline src/hooks/use-newsletter-subscription.js:89]: check sub status API HOOK 

@leonkoech remove the `console.log()`

File: src/hooks/use-newsletter.js [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+      method: "GET",
+      headers: {
+        "content-type": "application/json"
+      }
+    };
+
+    const data = await fetchUser({ config, authenticated: true });
+
+    if (data) {
+      console.log(data.active)
```
Comment (leonkoech) [inline src/hooks/use-newsletter.js:49]: @leonkoech remove the `console.log()`

File: src/modules/newsletters/button-module.js [hunk 12]
```diff
@@ -0,0 +1,4 @@
+export const BUTTON_TYPE_VALUES = ["primary_button","tertiary_button","secondary_button"]
```
Comment (leonkoech) [inline src/modules/newsletters/button-module.js:1]: to make it easier to create button components :)

Simply add one

File: src/modules/newsletters/text-module.js [hunk 13]
```diff
@@ -0,0 +1,4 @@
+export const USER_TYPE_VALUES = ["user_name","user_id","user_email","user_role"];
```
Comment (leonkoech) [inline src/modules/newsletters/text-module.js:1]: custom texts to insert into the newsletter

File: src/pages/newsletters/index.js [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+          event.target.header.value,
+           event.target.content.value,
+           email_List,
+           item_name
+          ):
+         preview_newsletter(
+            event.target.content.value,
+            is_HTML
+          ).then((val)=>{
+            console.log(val)
```
Comment (leonkoech) [inline src/pages/newsletters/index.js:56]: @leonkoech remove the `console.log()`

File: src/pages/newsletters/index.js [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+                  countEmails(new_list.length)
+
+                  selectRole(group_form_value)
+                  // countEmails(group_form_value.length)
+                  // listEmails(email_List[group_form_value])
+                }
+
+                // listEmails(new_list.length < 1?undefined:new_list)
+                // countEmails(new_list.length)
+                // console.log(group_form_value)
```
Comment (leonkoech) [inline src/pages/newsletters/index.js:177]: @leonkoech remove the comments

File: package.json (hunk 3 above)
Comment (gregv) [reply package.json:32]: Mmm toast

Comment (leonkoech) [reply package.json:32]: @gregv French toast is served! Bon appetit monsieur  🍞😂

File: package.json (hunk 2 above)
Comment (leonkoech) [reply package.json:22]: and edge cases I didn't anticipate but will test again.

File: src/components/newsletter/preview.js (hunk 6 above)
Comment (leonkoech) [reply src/components/newsletter/preview.js:11]: But again, this feature wasn't in the project description. The backend and the front end for the API request and the preview button were completed. Just some minor bug fixes and this feature can be implemented

File: src/hooks/use-newsletter-subscription.js (hunk 8 above)
Comment (leonkoech) [reply src/hooks/use-newsletter-subscription.js:39]: But tried a few things and got it working :))

File: src/components/newsletter/custom_button.js (hunk 5 above)
Comment (vertex) [reply src/components/newsletter/custom_button.js:30]: I'm @vertex here!  Thank you!!

**Review by gregv — CHANGES_REQUESTED** [review]
Asking for some small changes/questions to be answered

File: src/hooks/use-newsletter-subscription.js [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+
+          return error.message;
+      }
+  }, [getAccessTokenSilently]);
+
+  /**
+   * calls either the unsubscribe or subscribe based on the param
+   * @param {subscribe, user_id} strings
+   */
+  async function subscribe(subscription_state, user_id){
```
[the remark this answers is not in the document]
Comment (leonkoech) [reply src/hooks/use-newsletter-subscription.js:45]: It is a string of either `subscribe` or `unsubscribe` based on onClick event emitter from a button on either the subscribe or unsubscribe page which defines the path of the API call. 
eg api/newsletter-subs/unsubscribe/${user_id} or api/newsletter-subs/subscribe/${user_id}

This would mean that I wouldn't have to write the same function twice.

Is this an acceptable convention? @gregv

File: src/components/newsletter.js [hunk 17]
```diff
@@ -0,0 +1,14 @@
+import { useRouter } from "next/router";
+import NewsletterSubscriptionButton from "./buttons/newsletter-button";
+
+export default function NewsletterResponses(props) {
+  const router = useRouter();
+  const { user_Id } = router.query;
```
[the remark this answers is not in the document]
Comment (leonkoech) [reply src/components/newsletter.js:6]: I need some advice on this actually. 

Since the link on the email is something like /newsletters/unsubscribe/4564kj5623j4vdsfgsd, access to the last parameter which is the userID is needed. So I figured this was the best way to do it. 

Open to suggestions as to how I can achieve this. @gregv

File: src/hooks/use-newsletter.js [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+      return []
+    }
+  };
+
+  async function submit_email(subject, message,addresses, role) {
+    const email_data = {
+      body: message,
+      subject: subject,
+      addresses: addresses,
+      role: role
```
[the remark this answers is not in the document]
Comment (leonkoech) [reply src/hooks/use-newsletter.js:60]: I will add more comments everywhere on the next change of the PR to give ideas as to what functions do and what kind of parameters they expect as inputs

File: src/components/newsletter/selected_users.js [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+         }
+          props.subscribers_list(final_list)
+         }
+         
+        }
+       }
+
+     ></textarea>
+     :
+     <textarea
```
[the remark this answers is not in the document]
Comment (leonkoech) [reply src/components/newsletter/selected_users.js:51]: wow! will work on this with the rest of the team

File: src/pages/newsletters/subscribe/[user_Id].js [hunk 20]
```diff
@@ -0,0 +1,13 @@
+// import { useRouter } from "next/router";
+import NewsletterResponses from "../../../components/newsletter";
+export default function Subscribe_to_newsletter() {
+  // const router = useRouter();
+  // const { user_Id } = router.query;
```
[the remark this answers is not in the document]
Comment (leonkoech) [reply src/pages/newsletters/subscribe/[user_Id].js:5]: was part of the last tests. Removed it, thank you for catching :')

File: src/pages/newsletters/index.js [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
...21.355469 6 23 L 6 47 C 6 48.644531 7.355469 50 9 50 L 41 50 C 42.644531 50 44 48.644531 44 47 L 44 23 C 44 21.355469 42.644531 20 41 20 L 14.75 20 C 14.441406 19.007813 13.511719 16.074219 13.125 15 L 13.15625 15 C 11.519531 9.722656 14.5 4.109375 19.78125 2.46875 C 25.050781 0.832031 30.695313 3.796875 32.34375 9.0625 C 32.34375 9.066406 32.34375 9.089844 32.34375 9.09375 C 32.570313 9.886719 33.65625 13.40625 33.65625 13.40625 C 33.746094 13.765625 34.027344 14.050781 34.386719 14.136719 C 34.75 14.226563 35.128906 14.109375 35.375 13.832031 C 35.621094 13.550781 35.695313 13.160156 35.5625 12.8125 C 35.5625 12.8125 34.433594 9.171875 34.25 8.53125 L 34.25 8.5 C 32.78125 3.761719 28.601563 0.542969 23.9375 0.0625 C 23.550781 0.0234375 23.171875 0 22.78125 0 Z M 9 22 L 41 22 C 41.554688 22 42 22.445313 42 23 L 42 47 C 42 47.554688 41.554688 48 41 48 L 9 48 C 8.445313 48 8 47.554688 8 47 L 8 23 C 8 22.445313 8.445313 22 9 22 Z M 25 30 C 23.300781 30 22 31.300781 22 33 C 22 33.898438 22.398438 34.6875 23 35.1875 L 23 38 C 23 39.101563 23.898438 40 25 40 C 26.101563 40 27 39.101563 27 38 L 27 35.1875 C 27.601563 34.6875 28 33.898438 28 33 C 28 31.300781 26.699219 30 25 30 Z"></path>
```
[the remark this answers is not in the document]
Comment (leonkoech) [reply src/pages/newsletters/index.js:190]: the small lock Icons that provide feedback as to whether the text areas for the email list can be edited or not. 
Screenshot 2023-02-08 181500
Screenshot 2023-02-08 181446

Is a png better? or should I separate the SVG as a file or a component and import them? I'm not quite certain which one is better when it comes to balancing performance and maintainability @gregv

[the remark this answers is not in the document]
Comment (gregv) [reply src/pages/newsletters/index.js:190]: Use MaterialUI Icons, it will save your eyeballs and others from this sea of numbers
